### PR TITLE
i#3544 riscv64: Enable dynamorio unit test compile

### PR DESCRIPTION
**Pull request by shiptux**
Make unit tests compile by disabling them all.
Next, enable one test at a time when it passes.

Add RISCV-V 64 support to suite/tests/linux/infinite.c, and clients/drcachesim/tests/raw2trace_unit_tests.cpp
Notice: There are still a large number of unimplemented in Dynamorio RISC-V 64, and it cannot be tested via QEMU

Issue #3544

### CODE REVIEW

**Comment by shiptux** [comment]
@derekbruening

File: api/samples/CMakeLists.txt [hunk 1]
```diff
@@ -224,21 +224,25 @@ configure_DynamoRIO_global(OFF ON)
 
 # FIXME i#4372: Add functioning tests for samples that go beyond  # NON-PUBLIC
 # sanity checks.                                                  # NON-PUBLIC
-add_sample_client(bbbuf       "bbbuf.c"         "drmgr;drreg;drx")
-add_sample_client(bbcount     "bbcount.c"       "drmgr;drreg;drx")
-add_sample_client(bbsize      "bbsize.c"        "drmgr;drx")
+if (NOT RISCV64) # FIXME i#3544: Port tests to RISC-V64
```
Comment (derekbruening) [inline api/samples/CMakeLists.txt:227]: If the idea is to exclude absolutely everything in this file, could we have a single return() instead to avoid cluttering the file with if() statements?

File: api/samples/CMakeLists.txt [hunk 2]
```diff
@@ -224,21 +224,25 @@ configure_DynamoRIO_global(OFF ON)
...
 endif ()
-add_sample_client(div         "div.c"           "drmgr")
-add_sample_client(empty       "empty.c"         "")
-add_sample_client(memtrace_simple "memtrace_simple.c;utils.c" "drmgr;drreg;drutil;drx")
-add_sample_client(memval_simple   "memval_simple.c;utils.c"   "drmgr;drreg;drutil;drx")
-add_sample_client(instrace_simple "instrace_simple.c;utils.c" "drmgr;drreg;drx")
-add_sample_client(opcode_count    "opcode_count.cpp"    "drmgr;drreg;drx;droption")
+if (NOT RISCV64) # FIXME i#3544: Port tests to RISC-V64
+  add_sample_client(div         "div.c"           "drmgr")
+  add_sample_client(empty       "empty.c"         "")
```
Comment (derekbruening) [inline api/samples/CMakeLists.txt:240]: This empty client should definitely compile.

File: api/samples/CMakeLists.txt (hunk 1 above)
Comment (derekbruening) [inline api/samples/CMakeLists.txt:227]: nit: s/FIXME/TODO/ 

Ditto below.

(Really old code uses FIXME for everything but new code should only use it for very broken things: prefer TODO for missing not yet implemented pieces and XXX for optional changes.)

File: api/samples/CMakeLists.txt [hunk 3]
```diff
@@ -224,21 +224,25 @@ configure_DynamoRIO_global(OFF ON)
 
 # FIXME i#4372: Add functioning tests for samples that go beyond  # NON-PUBLIC
 # sanity checks.                                                  # NON-PUBLIC
-add_sample_client(bbbuf       "bbbuf.c"         "drmgr;drreg;drx")
-add_sample_client(bbcount     "bbcount.c"       "drmgr;drreg;drx")
-add_sample_client(bbsize      "bbsize.c"        "drmgr;drx")
+if (NOT RISCV64) # FIXME i#3544: Port tests to RISC-V64
+  add_sample_client(bbbuf       "bbbuf.c"         "drmgr;drreg;drx")
+  add_sample_client(bbcount     "bbcount.c"       "drmgr;drreg;drx")
+  add_sample_client(bbsize      "bbsize.c"        "drmgr;drx")
```
Comment (derekbruening) [inline api/samples/CMakeLists.txt:230]: These 3 should already compile: I don't see anything inside them that wouldn't compile for RISCV? What build error do you see with these?

File: core/ir/aarch64/codec.c [hunk 4]
```diff
@@ -6809,9 +6809,9 @@ decode_svemem_vec_imm5(uint enc, aarch64_reg_offset element_size, bool is_prefet
     default: ASSERT_NOT_REACHED();
     }
 
-    *opnd = opnd_create_vector_base_disp_aarch64(
-        zn, DR_REG_NULL, get_opnd_size_from_offset(element_size), DR_EXTEND_UXTX, false,
-        imm5, 0, mem_transfer, 0);
+    *opnd = opnd_create_vector_base_disp_aarch64(zn, DR_REG_NULL,
+                                                 get_opnd_size_from_offset(element_size),
+                                                 0, false, imm5, 0, mem_transfer, 0);
```
Comment (derekbruening) [inline core/ir/aarch64/codec.c:6814]: This change looks suspicious: please explain.

File: core/ir/aarch64/codec.c [hunk 5]
```diff
@@ -6824,13 +6824,6 @@ encode_svemem_vec_imm5(uint enc, aarch64_reg_offset element_size, bool is_prefet
         get_vector_element_reg_offset(opnd) != element_size)
         return false;
 
-    bool index_scaled;
-    uint index_scale_amount;
-    if (opnd_get_index_extend(opnd, &index_scaled, &index_scale_amount) !=
-            DR_EXTEND_UXTX ||
-        index_scaled || index_scale_amount != 0)
-        return false;
-
```
Comment (derekbruening) [inline core/ir/aarch64/codec.c:6833]: Again, removing code from the AArch64 decoder looks like a mistake?

File: core/ir/aarch64/instr_create_api.h [hunk 6]
```diff
@@ -10966,13 +10966,13 @@
  *             opnd_create_base_disp_aarch64(Rn, Rm, DR_EXTEND_UXTX, 0, 0, 0, OPSZ_1)
  *             For the  [\<Zn\>.S{, #\<imm\>}] variant:
  *             opnd_create_vector_base_disp_aarch64(Zn, DR_REG_NULL, OPSZ_4,
- *             DR_EXTEND_UXTX, 0, imm5, 0, opnd_size_from_bytes(dr_get_sve_vl() / 32), 0)
+ *             0, 0, imm5, 0, opnd_size_from_bytes(dr_get_sve_vl() / 32), 0)
```
Comment (derekbruening) [inline core/ir/aarch64/instr_create_api.h:10969]: Again, why is this RISCV PR deleting tokens from AArch64 decoder files?

File: suite/tests/common/getretaddr.c [hunk 7]
```diff
@@ -112,6 +112,9 @@ GLOBAL_LABEL(FUNCNAME:)
         mov      x0, x30
         ldr      x30, [sp], #16
         ret
+# elif defined(RISCV64)
+        /*# FIXME i#3544: Port tests to RISC-V64*/
```
Comment (derekbruening) [inline suite/tests/common/getretaddr.c:116]: Please use standard syntax: no #; end punctuation and a space before the closing *.  Ditto below.

File: suite/tests/tools.c [hunk 8]
```diff
@@ -564,6 +564,9 @@ GLOBAL_LABEL(code_self_mod:)
         sub      w1, w1, #1
         cbnz     w1, repeat1
         ret
+#elif defined(RISCV64)
+        /*FIXME i#3544: Port tests to RISC-V64*/
```
Comment (derekbruening) [inline suite/tests/tools.c:568]: Please add spaces and end punctuation.  Ditto below.

File: suite/tests/tools.c [hunk 9]
```diff
@@ -597,32 +600,37 @@ GLOBAL_LABEL(FUNCNAME:)
...
 #undef FUNCNAME
 #define FUNCNAME code_inc
         DECLARE_FUNC(FUNCNAME)
 GLOBAL_LABEL(FUNCNAME:)
+#if !defined(RISCV64) /*FIXME i#3544: Port tests to RISC-V64*/
         mov      REG_SCRATCH0, ARG1
         INC(REG_SCRATCH0)
         RETURN
         END_FUNC(FUNCNAME)
+#endif
```
Comment (derekbruening) [inline suite/tests/tools.c:612]: It seems incorrect to exclude END_FUNC when DECLARE_FUNC is included.

File: core/ir/aarch64/codec.c (hunk 5 above)
Comment (shiptux) [reply core/ir/aarch64/codec.c:6833]: Ditto below.

File: core/ir/aarch64/instr_create_api.h (hunk 6 above)
Comment (shiptux) [reply core/ir/aarch64/instr_create_api.h:10969]: Ditto below.

**Comment by shiptux** [comment]
I updated this PR to the latest master branch the day before yesterday. And then the CI won't be pasted. So I reverted the updated commit, this is the reason why some aarch64 decoder modifications in my PR. @derekbruening

**Comment by shiptux** [comment]
I updated this PR to the latest master branch, and then I found that the CI could not pass. So I revert the updated commit, this is the reason why my PR has some modifications of the aarch64 decoder. @derekbruening

File: core/ir/aarch64/codec.c (hunk 4 above)
Comment (derekbruening) [reply core/ir/aarch64/codec.c]: I don't understand: you reverted an AArch64 commit in this PR?  Then that is a change from your PR (and likely a change we do not want to merge).

In the old days Github would not hide changes from merges: so a reviewer looking at what changed since the last review would see master branch changes if the author did an update mid-review.  If that is what is happening (because it is a personal repo?) then please do not update this PR branch until the review is finished.

File: suite/tests/CMakeLists.txt [hunk 10]
```diff
@@ -1997,7 +2004,9 @@ if (X86)
 elseif (AARCHXX)
   set(control_flags "nzcv")
 endif ()
+if (NOT RISCV64) # TODO i#3544: Port tests to RISC-V 64
 tobuild(common.${control_flags} common/${control_flags}.c)
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt:2008]: Missing indentation

File: suite/tests/CMakeLists.txt [hunk 11]
```diff
@@ -2782,11 +2811,14 @@ if (X86)
...
 endif ()
 
-tobuild_appdll(client.drwrap-test client-interface/drwrap-test.c)
-get_target_path_for_execution(drwrap_libpath client.drwrap-test.appdll "${location_suffix}")
-tobuild_ci(client.drwrap-test client-interface/drwrap-test.c "" "" "${drwrap_libpath}")
-use_DynamoRIO_extension(client.drwrap-test.dll drwrap)
-tochcon(client.drwrap-test.appdll textrel_shlib_t)
+if (NOT RISCV64) # TODO i#3544: Port tests to RISC-V 64
+  tobuild_appdll(client.drwrap-test client-interface/drwrap-test.c)
+  get_target_path_for_execution(drwrap_libpath client.drwrap-test.appdll "${location_suffix}")
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt:2816]: Line too long: please wrap

File: suite/tests/CMakeLists.txt [hunk 12]
```diff
@@ -3574,17 +3613,19 @@ if (BUILD_CLIENTS)
       endif ()
     endif ()
 
+    if (NOT RISCV64)
     # Test offline traces.
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt:3617]: Missing indentation on these 4 comment lines

File: suite/tests/CMakeLists.txt [hunk 13]
```diff
@@ -3810,7 +3851,7 @@ if (BUILD_CLIENTS)
         "@-simulator_type@basic_counts" "")
       unset(tool.drcacheoff.basic_counts_rawtemp)
     endif ()
-
+    if (NOT RISCV64)
     torunonly_drcacheoff(opcode_mix ${ci_shared_app} ""
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt:3855]: Missing indentation on all these lines below, and missing TODO comment

File: suite/tests/CMakeLists.txt [hunk 14]
```diff
@@ -3574,17 +3613,19 @@ if (BUILD_CLIENTS)
...
-
+    # TODO i#3544: Port tests to RISC-V 64
+      get_target_path_for_execution(drcachesim_path drcachesim "${location_suffix}")
+      prefix_cmd_if_necessary(drcachesim_path ON ${drcachesim_path})
+      get_target_path_for_execution(drraw2trace_path drraw2trace "${location_suffix}")
+      prefix_cmd_if_necessary(drraw2trace_path ON ${drraw2trace_path})
+
+      torunonly_drcachesim(opcode_mix ${ci_shared_app}
+        "-instr_encodings -simulator_type opcode_mix" "")
+    endif(NOT RISCV64)
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt:3628]: nit: missing space after endif

File: suite/tests/CMakeLists.txt [hunk 15]
```diff
@@ -4107,7 +4150,6 @@ if (BUILD_CLIENTS)
     if (AARCH64)
       torunonly_drcachesim(drstatecmp-fuzz drstatecmp-fuzz-app
         "-trace_after_instrs 10M -enable_drstatecmp" "" "")
-      set(tool.drcachesim.drstatecmp-fuzz_timeout 180)
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt:4110]: This looks like a mistake: why was this line deleted?

File: suite/tests/CMakeLists.txt [hunk 16]
```diff
@@ -4650,16 +4698,21 @@ if (UNIX)
     endif ()
   endif ()
 
+  if (NOT RISCV64)
   # We pass -native_exec_list which causes us to call strcasecmp in libc,
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt:4702]: Missing indentation on these lines below

File: suite/tests/CMakeLists.txt [hunk 17]
```diff
@@ -4680,22 +4734,28 @@ if (UNIX)
...
-  if (NOT APPLE)
+  if (NOT APPLE AND NOT RISCV64)
     # FIXME i#125: bugs in these, were disabled in the past
+    # TODO i#3544: Port tests to RISC-V 64
     tobuild_ops(linux.vfork linux/vfork.c "" "${execve-sub-path}")
     #tobuild(linux.vfork-fib linux/vfork-fib.c)
   endif ()
 
+  if (NOT RISCV64)
   # runs of other builds with custom DR options
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt:4752]: Missing indentation on these 2 lines

**Comment by derekbruening** [comment]
> I updated this PR to the latest master branch, and then I found that the CI could not pass. So I revert the updated commit, this is the reason why my PR has some modifications of the aarch64 decoder

Please add comments/labels explaining that in the future as otherwise it looks like an unintentional mistake when looking just at the diff.

File: core/ir/aarch64/codec.c (hunk 4 above)
Comment (derekbruening) [reply core/ir/aarch64/codec.c]: Looks like github is successfully hiding merge-from-master changes -- so this was all about an explicitly reverted commit; see below about clearly marking that and indicating what the plan is: only temporary to test the CI or actually proposing to permanently revert (in which case it should be a separate PR and should bring in the authors of the being-reverted PR).

**Comment by derekbruening** [comment]
Looks good, except it looks like you've missed one comment about the too-long line: once that's fixed we can merge it in.

**Comment by shiptux** [comment]
> Looks good, except it looks like you've missed one comment about the too-long line: once that's fixed we can merge it in.
 
I don't know if you said the too-long line is:
`if (NOT RISCV64) # TODO i#3544: Port tests to RISC-V 64`
Or
`get_target_path_for_execution(drwrap_libpath client.drwrap-test.appdll "${location_suffix}")`
I have already moved `# TODO i#3544: Port tests to RISC-V 64` to the second line.
```
if (NOT RISCV64) # TODO i#3544: Port tests to RISC-V 64
if (NOT RISCV64)
  # TODO i#3544: Port tests to RISC-V 64
  tobuild_appdll(client.drwrap-test client-interface/drwrap-test.c)
  get_target_path_for_execution(drwrap_libpath client.drwrap-test.appdll "${location_suffix}")
```

File: suite/tests/CMakeLists.txt (hunk 11 above)
Comment (derekbruening) [reply suite/tests/CMakeLists.txt:2817]: > I don't know if the too-long line you said that is:
> if (NOT RISCV64) # TODO i#3544: Port tests to RISC-V 64
> Or
> get_target_path_for_execution(drwrap_libpath client.drwrap-test.appdll "${location_suffix}")
> I have already moved # TODO i#3544: Port tests to RISC-V 64 to the second line.
> 
> -if (NOT RISCV64) # TODO i#3544: Port tests to RISC-V 64
> +if (NOT RISCV64)
> + # TODO i#3544: Port tests to RISC-V 64
>   tobuild_appdll(client.drwrap-test client-interface/drwrap-test.c)
>   get_target_path_for_execution(drwrap_libpath client.drwrap-test.appdll "${location_suffix}")

The comment is always immediately below the referenced line.  It is the get_target_path... which is over 80 chars.

**Comment by shiptux** [comment]
Sorry for replying too late. These changes are all done. @derekbruening

**Review by derekbruening — APPROVED** [review]
